integration/activemq-kafka/activemq-kafka-bridge/src/main/java/org/apache/activemq/artemis/integration/kafka/bridge/KafkaProducerBridge.java

```diff
@@ -0,0 +1,484 @@
+         }
+         this.queue = (Queue) b.getBindable();
+
+         this.kafkaProducer = kafkaProducerFactory.create(kafkaConfig(configuration));
+
+         List<PartitionInfo> topicPartitions = kafkaProducer.partitionsFor(topicName);
+         if (topicPartitions == null || topicPartitions.size() == 0) {
+            throw new Exception(connectorName + ": topic " + topicName + " not found");
+         }
+
+         this.retryCount.set(0);
+         this.isStarted.set(true);
+         connect();
+         ActiveMQKafkaLogger.LOGGER.bridgeStarted(connectorName);
+      }
+   }
+
+   public void connect() throws Exception {
+      synchronized (this) {
+         if (!isConnected && isStarted.get()) {
+            isConnected = true;
+            this.queue.addConsumer(this);
+            this.queue.deliverAsync();
+            ActiveMQKafkaLogger.LOGGER.bridgeConnected(connectorName);
+         }
+      }
+   }
+
+   @Override
+   public void disconnect() {
+      synchronized (this) {
+         if (isConnected) {
+            if (queue != null) {
+               this.queue.removeConsumer(this);
+            }
+
+            cancelRefs();
+            if (queue != null) {
+               queue.deliverAsync();
+            }
+            isConnected = false;
+            ActiveMQKafkaLogger.LOGGER.bridgeDisconnected(connectorName);
+         }
+      }
+   }
+
+   @Override
+   public void stop() {
+      synchronized (this) {
+         if (!this.isStarted.get()) {
+            return;
+         }
+         ActiveMQKafkaLogger.LOGGER.bridgeReceivedStopRequest(connectorName);
+
+         disconnect();
+
+         kafkaProducer.close();
+         kafkaProducer = null;
+         this.isStarted.set(false);
+         ActiveMQKafkaLogger.LOGGER.bridgeStopped(connectorName);
+      }
+   }
+
+   @Override
+   public boolean isStarted() {
+      return this.isStarted.get();
+   }
+
+   @Override
+   public String getName() {
+      return this.connectorName;
+   }
+
+   @Override
+   public boolean supportsDirectDelivery() {
+      return false;
+   }
+
+   @Override
+   public HandleStatus handle(MessageReference ref) throws Exception {
+      if (filter != null && !filter.match(ref.getMessage())) {
+         return HandleStatus.NO_MATCH;
+      }
+
+      synchronized (this) {
+         ref.handled();
+
+         Message message = ref.getMessage();
+
+         synchronized (refs) {
+            refs.put(ref.getMessage().getMessageID(), ref);
+         }
+
+         Integer partition = null;
+         SimpleString groupdID = message.getGroupID();
+         if (groupdID != null) {
+            List partitions = kafkaProducer.partitionsFor(topicName);
+            int numPartitions = partitions.size();
+            partition = Utils.toPositive(Utils.murmur2(groupdID.getData())) % numPartitions;
+         }
+
+         //Compaction/Record Key (equivalent to Artemis LVQ)
+         SimpleString key = message.getLastValueProperty();
```

Comment:
I'm not an expert on Artemis but can you explain me why the LVQ is used as key for partitions ?

Reply:
Its not used for partitions, Key in Kafka is used for compaction. We use groupId (JMS MessageGroup) to sort partition see code above.